rustls/src/conn.rs

```diff
@@ -1080,17 +1086,29 @@ impl ConnectionCommon {
     }
 }
 
-pub(crate) trait HandleState: Sized {
-    type Data;
-    type Config;
-
+pub(crate) trait State<Data, Config>: Send + Sync {
     fn handle(
-        self,
+        self: Box<Self>,
+        cx: &mut Context<'_, Data, Config>,
         message: Message,
-        data: &mut Self::Data,
-        common: &mut ConnectionCommon,
-        config: &Self::Config,
-    ) -> Result<Self, Error>;
+    ) -> Result<Box<dyn State<Data, Config>>, Error>;
+
+    fn export_keying_material(
+        &self,
+        _output: &mut [u8],
+        _label: &[u8],
+        _context: Option<&[u8]>,
+    ) -> Result<(), Error> {
+        Err(Error::HandshakeNotComplete)
+    }
+
+    fn perhaps_write_key_update(&mut self, _common: &mut ConnectionCommon) {}
+}
+
+pub(crate) struct Context<'a, Data, Config> {
+    pub(crate) common: &'a mut ConnectionCommon,
+    pub(crate) data: &'a mut Data,
+    pub(crate) config: &'a Config,
```

Comment:
`config` will go away after #681, right?

Then couldn't we just add `data: Data` to `ConnectionCommon` and rename `ConnectionCommon` to `Connection`, and avoid having `Context` at all? In that case we'd have `trait State<Data>`?

Alternatively, if we merge this PR, then do we need #681 at all?

I'm guessing we need #681 because after the handshake we'll have a `State` to indicate we're done handshaking but we don't want to have a `Config` to pass to it.

Reply:
I do definitely think #681 is still useful. In the future I think we don't want to pass all of `ConnectionCommon` into the states as part of the state machines, but only the more limited parts that are required (writing out some of the state for use by the API, and sending messages). In particular, if we want to do more borrowing of incoming message payloads, we cannot pass a mutable reference to the buffers that those messages reside in. In my current thinking, we might split `ConnectionCommon` into `Incoming`, `Outgoing`, and `CommonState` types, and then pass only `Data`, `Outgoing`, and `CommonState` into the handshake.

I did an earlier attempt at the state machine refactoring where `ConnectionCommon` grew type arguments for `Data` and `Config` directly, but this didn't work out well. I don't remember exactly why or how, though it might have been related to how the `ClientSession` gets initialized.

Reply:
I guess it could still make sense to have call `CommonState` `ConnectionState` instead and add the `Data` argument to it, and then it could live inside the `ConnectionCommon` which could be renamed to `Connection`. Still, I think the important part in this PR is to flatten the `HandleState` and `client::hs::State`/`server::hs::State` traits together, which we should still do anyway.